Indication of Mutation types onto graph

joswinter opened the pull request:
Indicates mutation types by first generating a unified coordinate system, a reference coordinate system, empty nodes between gap edges and looping over the vertices to determine the mutation type if it has one.

joswinter committed:
Rebuild and renamed coordinates in Segment

start, end = coordinate system in file
unified = unified coordinate system with all sequences
reference = coorindate system in comparision with reference sequence

joswinter committed:
Added Graph.copy method

joswinter committed:
Added UnifiedPositionTraverser

Maps all segments to a unified coordinate system.

joswinter committed:
Added EmptySegmentTraverser

Divide gap edges into empty nodes.

joswinter committed:
Added ReferencePositionTraverser

Maps all segments to the reference coordinate system.

joswinter committed:
Added MutationIndicationTraverser

Indicates mutations onto a SequenceSegmentGraph

joswinter committed:
Added missing package-info

joswinter committed:
Added GraphAdapter.copy test

joswinter committed:
Renamed variable names in test

Confusing previous class names.

joswinter committed:
Added UnifiedPositionTraverserTest

jorenham commented on src/main/java/nl/tudelft/lifetiles/graph/controller/GraphController.java:
```diff
@@ -54,7 +59,16 @@ public final void initialize(final URL location,
                 "data/test_set/simple_graph", gf);
-        return gr;
+        Sequence reference = gr.getSources().iterator().next().getSources()
+                .iterator().next();
```
Creating all those new traversers is blasphemy according to the dependency-injection design pattern. You should either use a factory or pass the traversers as abstract type to the method.

jorenham commented on src/main/java/nl/tudelft/lifetiles/graph/models/jgrapht/JGraphTGraphAdapter.java:
```diff
@@ -223,4 +223,22 @@ private void removeEdge(final Edge<V> edge) {
+     */
+    @Override
+    public final Graph<V> copy() {
+        Graph<V> graph = new JGraphTGraphAdapter<V>(edgeFact);
```
Dependency-injection

jorenham commented on src/main/java/nl/tudelft/lifetiles/graph/view/TileView.java:
```diff
@@ -171,9 +171,9 @@ private boolean segmentFree(final int ind, final SequenceSegment segment) {
         for (int height = 0; height < segment.getSources().size(); height++) {
             if (index + height < lanes.size()) {
-                lanes.set(index + height, segment.getEnd());
+                lanes.set(index + height, segment.getUnifiedEnd());
```
Double code, you should make `index + height, segment.getUnifiedEnd()` a variable outside the condition scope.

jorenham commented on src/main/java/nl/tudelft/lifetiles/traverser/models/EmptySegmentTraverser.java:
```diff
@@ -0,0 +1,185 @@
+     * Traverses the copied graph.
+     */
+    public final void traverseGraph() {
+        for (SequenceSegment vertex : new HashSet<SequenceSegment>(
```
Creating a `List` is quicker than creating a `Set` because of the uniqueness checks (although `O(1)`) and creating a set here does not have additional value in comparison to creating a list.

jorenham commented on src/main/java/nl/tudelft/lifetiles/traverser/models/EmptySegmentTraverser.java:
```diff
@@ -0,0 +1,185 @@
+        while (!it.isEmpty()) {
+            SortedEdge edge = it.poll();
+            SequenceSegment destination = edge.getSegment();
+            HashSet<Sequence> sources = new HashSet<Sequence>(
```
Use abstract type.

jorenham commented on src/main/java/nl/tudelft/lifetiles/traverser/models/MutationIndicationTraverser.java:
```diff
@@ -0,0 +1,82 @@
+     */
+    private void traverseVertex(final SequenceSegment vertex) {
+        if (!vertex.getSources().contains(referenceVar)) {
+            if (vertex.getContent() instanceof SegmentEmpty) {
```
Double code (`vertex.setMutation`). You should do this after the condition block and inside there, just select the mutation.

joswinter committed:
Added graph traversals to Tile

jorenham commented on src/main/java/nl/tudelft/lifetiles/traverser/models/ReferencePositionTraverser.java:
```diff
@@ -0,0 +1,129 @@
+     */
+    public final Graph<SequenceSegment> traverseGraph(
+            final Graph<SequenceSegment> graph) {
+        graphVar = graph.copy();
```
The graph should not be copied inside the constructor but before the call to it.

jorenham commented on src/main/java/nl/tudelft/lifetiles/traverser/models/EmptySegmentTraverser.java:
```diff
@@ -0,0 +1,185 @@
+     */
+    private void traverseVertex(final SequenceSegment vertex) {
+        HashSet<Sequence> buffer = new HashSet<Sequence>(vertex.getSources());
+        PriorityQueue<SortedEdge> it = getSortedEdges(graphVar
```
`Queue` as type.

joswinter committed:
Removed double code TileView

Double code, you should make index + height, segment.getUnifiedEnd() a
variable outside the condition scope.

joswinter committed:
Program to Interface in EmptySegmentTraverser

Use abstract type.

joswinter committed:
Added trailing end line files to tests

Missing trailing newline

joswinter committed:
Removed internal Graph copy in traversers

The graph should not be copied inside the constructor but before the
call to it.

joswinter committed:
EmptySegmentTraverser: copy to list instead of set

Creating a List is quicker than creating a Set because of the uniqueness
checks (although O(1)) and creating a set here does not have additional
value in comparison to creating a list.

joswinter committed:
Fixed long line in TileView

[ERROR] src\main\java\nl\tudelft\lifetiles\graph\view\TileView.java[153]
(sizes) LineLength: Line is longer than 80 characters (found 81).

joswinter committed:
MutationTraverser removed double code.

Double code (vertex.setMutation). You should do this after the condition
block and inside there, just select the mutation.

joswinter committed:
Implemented DI into Graph.copy

Dependency-injection

rutgervandenberg commented on src/main/java/nl/tudelft/lifetiles/traverser/models/EmptySegmentTraverser.java:
```diff
@@ -0,0 +1,186 @@
+    /**
+     * Traverses the graph.
+     */
+    public final void traverseGraph() {
```
Should rename this to something more descriptive. 'traverse' does not tell me what it will do. Neither is it explained in the corresponding javadoc.

rutgervandenberg commented on src/main/java/nl/tudelft/lifetiles/traverser/models/EmptySegmentTraverser.java:
```diff
@@ -0,0 +1,186 @@
+     *            the edges to sort.
+     * @return priority queue of sorted edges.
+     */
+    @Deprecated
```
No longer needed after #91. Should probably wait for it.

rutgervandenberg commented on src/main/java/nl/tudelft/lifetiles/traverser/models/MutationIndicationTraverser.java:
```diff
@@ -0,0 +1,84 @@
+     *            The graph to traverse.
+     * @return the traversed graph.
+     */
+    public final Graph<SequenceSegment> traverseGraph(
```
Should rename this to be more descriptive.

rutgervandenberg commented on src/main/java/nl/tudelft/lifetiles/traverser/models/ReferencePositionTraverser.java:
```diff
@@ -0,0 +1,129 @@
+     *            Graph to be traversed.
+     * @return Traversed graph with reference positions.
+     */
+    public final Graph<SequenceSegment> traverseGraph(
```
Should rename this to be more descriptive.

rutgervandenberg commented on src/main/java/nl/tudelft/lifetiles/traverser/models/UnifiedPositionTraverser.java:
```diff
@@ -0,0 +1,65 @@
+     *            Graph to be traversed, calculate unified positions.
+     * @return traversed copy of the graph.
+     */
+    public final Graph<SequenceSegment> traverseGraph(
```
Should rename this to be more descriptive. Basically, you removed the graphtraverser Interface, but you didn't actually change the implementations. They're still implicitly following the same interface.

joswinter replied:
I reintroduced the Interface which is needed for the TraverserFactory. I added better description in the javadoc. And changed the internal traverseGraph names to more descriptive ones.

joswinter committed:
Added interface Traverser for TraverserFactory

joswinter committed:
Implemented TraverserFactory

joswinter committed:
Tests and Tile should use TraverserFactory

jorenham commented on src/main/java/nl/tudelft/lifetiles/traverser/models/MutationIndicationTraverser.java:
```diff
@@ -0,0 +1,84 @@
+    private void traverseVertex(final SequenceSegment vertex) {
+        if (!vertex.getSources().contains(referenceVar)) {
+            Mutation mutation;
+            if (vertex.getContent() instanceof SegmentEmpty) {
```
You could avoid using this `if`-`else` construct with an `instanceof` (code smell) by using polymorphism (don't confuse it with the mutation type). See http://stackoverflow.com/a/519537/1627479 for an example.

joswinter committed:
Added better comments to traversers

Should rename this to be more descriptive.

joswinter committed:
Traverser better internal methodNames

joswinter committed:
Added isEmpty to segmentContent

You could avoid using this if-else construct with an instanceof (code
smell) by using polymorphism (don't confuse it with the mutation type).
See http://stackoverflow.com/a/519537/1627479 for an example.

joswinter committed:
Merge remote-tracking branch 'ProgrammingLife5/dev' into graphViewer---showcase

Conflicts:
	src/main/java/nl/tudelft/lifetiles/graph/models/jgrapht/JGraphTGraphAdapter.java
	src/main/java/nl/tudelft/lifetiles/graph/models/sequence/SequenceSegment.java

joswinter committed:
Fixed UnifiedPositionTraverser to use temporary map

Used to avoid concurrent modification exceptions.

joswinter committed:
EmptySegmentTraverser uses temp map + sorted edges

Doesn't work because of bug in the graph, if the vertex is changed it
can't be found in the graph.

joswinter committed:
Remove unified and mutation in segment.hashcode

Don't know if equals and compareTo should be changed. But compareTo
can't be changed.

joswinter committed:
moved calculateMutation to sequence segment

joswinter committed:
changed to determineMutation

calculate -> determineMutation

joswinter committed:
Removed TraverserInterface and TraverserFactory
